feat: Show applicable formats/classes in image/link WYSIWYG dialog

nadirayasmeen opened the pull request:
Related issue: hannonhill/Cascade#3915

Commit:
Check if the format applies to the current selection before adding it to class list in image/link plugins #3915

Commit:
Removes commented out code #3915

Commit:
Only include classes dropdown that are also in the image or link class list #3915

nadirayasmeen commented on src/plugins/link/src/main/js/ui/Dialog.js:
```diff
@@ -383,10 +384,25 @@ define(
+
+        if (classList) {
+          // Add a 'None' option to the beginning if it is not already present.
+          if (typeof classList[0] !== 'object') {
```
I kept trying to extract this out by using `CascadeUtils.appendOptionsToDropDown(classList, data);` (same as in the image plugin), but `grunt` kept throwing syntax errors for some reason when I tried to compile the change. I gave up temporarily to keep the issue moving 😞

rgriffith replied:
We worked through this with a "common plugin"

nadirayasmeen commented on src/plugins/cascade/src/main/js/core/Utils.js:
```diff
@@ -0,0 +1,67 @@
+ *
+ */
+define(
+  'tinymce.plugins.cascade.core.Utils',
```
Bolt configuration is pretty particular about the file hierarchy format...

nadirayasmeen replied:
This `Utils` plugin doesn't get compiled and resulting files copied over to Cascade like for `image` & `link` & seems to work. I'm assuming that this plugin, like the regular `Settings` and `Utils` in the image plugin, just get compiled in the image & link plugins.

rgriffith replied:
Yeah, I think this is fine. I recall looking through WordPress' customizations at one point and they have a similar setup with some specific plugins.

Commit:
temp: Updates logic for populating class dropdown #3915

nadirayasmeen commented on src/plugins/cascade/src/main/js/core/Utils.js:
```diff
@@ -9,36 +9,95 @@ define(
+
+        if (nodeName.toLowerCase() !== element) {
+          var elementNode = editor.dom.create(element, {});
+         // editor.selection.setNode(elementNode); //adds node to DOM
```
Create didn't insert a node into the editor, but also wouldn't work as a `Node` parameter to [dom.is()](https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.domutils/#is) unless it's either appended to the current selection, or attempted to set the node in the selection object: `selection.setNode(newNode)` but those options insert the element (which we're trying to avoid)

rgriffith replied:
We worked through this by using the Formatter class and an empty element on creation.

nadirayasmeen commented on src/plugins/cascade/src/main/js/core/Utils.js:
```diff
@@ -9,36 +9,95 @@ define(
       };
 
+      // Attempts to checks the selector up until the element since it hasn't been inserted into the editor yet
+      var matchesSelectorForNewElement = function (element, selectedNode, selector, editor) {
```
I moved the logic added in the earlier `wysiwyg.js` commits to the plugin side & updated it to so [dom.is](https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.domutils/#is) could semi-work against the selector to help determine if a format would apply to the new element: `matchesSelectorForNewElement` However, it's definitely imperfect & for more complex selectors, those classes will not get added to the dropdown.

rgriffith replied:
We ended up reworking this such that we test if the format is applicable to a node which mimics the functionality of TinyMCE's Format dropdown.

nadirayasmeen commented:
Was able to get the class populating correctly for existing links/images using dom.is: https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.domutils/#is

[formatter.matchNode](https://www.tiny.cloud/docs/api/tinymce/tinymce.formatter/#matchnode) was always return `false` for new/existing elements, but I think I was misunderstanding how it works. If it only returns `true` once the format has been applied to the element, then that makes sense, since the format hasn't been applied yet. So either `dom.is()`, `formatter.canApply()` seems to work in this case.

However, still having some trouble getting the class dropdown populating correctly when inserting a new image/link though; not really sure how to proceed or if I'm possibly missing something available in the TinyMCE API that would help?

Commit:
temp: Rework logic to return custom formats instead of classes of those formats #3915

rgriffith committed:
Allow multiple applicable styles to be selected within image and link plugin dialogs
hannonhill/Cascade#3915

Commit:
Adds option groups, convert simple classes to use multi-select, logic for applying styles-only formats #3915

Adds option groups when an advanced format contains both custom formats & existing classes
Custom formats that do not include classes only in-line styles weren't getting styles applied to the elements
Simple class format now use multi-select instead of dropdown

nadirayasmeen commented on src/plugins/cascade/src/main/js/core/CustomStyleFormatsUtils.js:
```diff
@@ -112,11 +226,39 @@ define(
         return uniqueImageClasses;
       };
 
+      var getFormatInlineStyles = function (selectedCustomFormatNames, customStyleFormats) {
```
I originally added logic for merging inline styles with any existing inline styles but ran into a lot of issues getting it working properly & the logic was getting more complex & time-consuming on top of what was already added. 

Mainly, it was hard to determine if an inline style was existing previously because for some styles, i.e. something like `border: 1px solid green` would automatically get converted into a separate style, 'border-width: 1px'; when the dialog was submitted; so simply comparing to the format "combined" version wouldn't work. Wondering if there's a more convenient way to do this, or possibly tabling it for now & considering in post issue (hopefully, there isn't a heavy reliance on inline styling).

rgriffith replied:
This was reworked in 270813b

rgriffith committed:
Ensure class select options have a value #3915

rgriffith committed:
General code cleanup and reuse #3915

rgriffith committed:
Add missing type to list items
hannonhill/Cascade#3915

rgriffith committed:
Refactor merging of selected formats and classes
hannonhill/Cascade#3915

rgriffith committed:
Pass style attribute into link
hannonhill/Cascade#3915

rgriffith committed:
Always output opt groups
hannonhill/Cascade#3915

rgriffith committed:
Sort options when building optgroups hannonhill/Cascade#3915

rgriffith committed:
Reworking logic and generation of format and class list items.
hannonhill/Cascade#3915

rgriffith committed:
Pass editor settings not editor when getting link class list setting.
hannonhill/Cascade#3915

rgriffith committed:
Update handling of existing styles and styles applied through formats
hannonhill/Cascade#3915

rgriffith committed:
Inline styles applied via a Format take priority over manually added styles
Prevents issue where a customized inline style would prevent the Format from being selected on subsequent dialog loading, since the style no longer matches the Format.
hannonhill/Cascade#3915

rgriffith committed:
Fix misspelled method name
hannonhill/Cascade#3915

Commit:
Trims existing classes before splitting them hannonhill/Cascade#3915

rgriffith committed:
Code cleanup and restructuring
hannonhill/Cascade#3915

nadirayasmeen commented:
Code restructuring looks good. Good call moving Inline styles logic to its own util 👍

rgriffith reviewed (approved):
Majority of changes are within this PR, which @nadirayasmeen and I have been cleaning up as best as possible without refactoring too much of the existing plugin code to prevent potentially breaking something.

Commit:
Adds additional checks so image/link dialogs aren't broken in non-editor configuration WYSIWYGs hannonhill/Cascade#3915

Commit:
Update Format control label, add help text for selecting multiple formats hannonhill/Cascade#3915

Commit:
Fixes issue where css class option group gets labeled as 'Custom Formats' hannonhill/Cascade#3915

Commit:
Fixes issue where new elements didn't load simple classes in dialogs hannonhill/Cascade#3915

Commit:
Update format/class label to "Styling" hannonhill/Cascade#3915
